File: packages/node/runtime/src/frames.js
import { WreError, ErrorKind } from "./errors.js";

const HEADER_BYTES = 8;
const MAX_JSON_BYTES = 64 * 1024 * 1024;
const MAX_BIN_BYTES = 512 * 1024 * 1024;

export function encodeFrame(jsonValue, bin) {
  const json = Buffer.from(JSON.stringify(jsonValue), "utf8");
  const binBuf = bin === undefined || bin === null ? Buffer.alloc(0) : Buffer.from(bin);
  const header = Buffer.alloc(HEADER_BYTES);
  header.writeUInt32BE(json.length, 0);
  header.writeUInt32BE(binBuf.length, 4);
  return Buffer.concat([header, json, binBuf]);
}

export class FrameDecoder {
  constructor() {
    this.buffer = Buffer.alloc(0);
  }

  push(chunk) {
    const incoming = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
    this.buffer = this.buffer.length === 0 ? incoming : Buffer.concat([this.buffer, incoming]);

    const frames = [];
    for (;;) {
      if (this.buffer.length < HEADER_BYTES) break;
      const jsonLen = this.buffer.readUInt32BE(0);
      const binLen = this.buffer.readUInt32BE(4);
      if (jsonLen > MAX_JSON_BYTES) {
        throw new WreError(ErrorKind.Protocol, `frame json length ${jsonLen} exceeds the 64 MiB cap`);
      }
      if (binLen > MAX_BIN_BYTES) {
        throw new WreError(ErrorKind.Protocol, `frame binary length ${binLen} exceeds the 512 MiB cap`);
      }
      const total = HEADER_BYTES + jsonLen + binLen;
      if (this.buffer.length < total) break;

      const jsonBytes = this.buffer.subarray(HEADER_BYTES, HEADER_BYTES + jsonLen);
      const binBytes = this.buffer.subarray(HEADER_BYTES + jsonLen, total);
      const json = JSON.parse(jsonBytes.toString("utf8"));
      const bin = binLen === 0 ? Buffer.alloc(0) : Buffer.from(binBytes);
      frames.push({ json, bin });

      this.buffer = this.buffer.subarray(total);
    }
    return frames;
  }
}

File: packages/python/wre_runtime/frames.py
from __future__ import annotations

import json
import struct
from typing import Any, BinaryIO, Dict, List, Optional, Tuple

from .errors import ProtocolError

_HEADER = struct.Struct(">II")
_MAX_JSON_LEN = 64 * 1024 * 1024
_MAX_BIN_LEN = 512 * 1024 * 1024


def encode_frame(envelope: Dict[str, Any], bin_part: bytes = b"") -> bytes:
    json_bytes = json.dumps(envelope, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    return _HEADER.pack(len(json_bytes), len(bin_part)) + json_bytes + bin_part


class FrameDecoder:
    def __init__(self) -> None:
        self._buffer = bytearray()

    def feed(self, chunk: bytes) -> List[Tuple[Dict[str, Any], bytes]]:
        self._buffer.extend(chunk)
        frames: List[Tuple[Dict[str, Any], bytes]] = []
        while True:
            if len(self._buffer) < _HEADER.size:
                break
            json_len, bin_len = _HEADER.unpack_from(self._buffer, 0)
            _check_caps(json_len, bin_len)
            total = _HEADER.size + json_len + bin_len
            if len(self._buffer) < total:
                break
            json_bytes = bytes(self._buffer[_HEADER.size : _HEADER.size + json_len])
            bin_bytes = bytes(self._buffer[_HEADER.size + json_len : total])
            del self._buffer[:total]
            envelope = json.loads(json_bytes.decode("utf-8"))
            frames.append((envelope, bin_bytes))
        return frames


def read_frame(stream: BinaryIO) -> Optional[Tuple[Dict[str, Any], bytes]]:
    header = _read_exact(stream, _HEADER.size, allow_eof_at_start=True)
    if header is None:
        return None
    json_len, bin_len = _HEADER.unpack(header)
    _check_caps(json_len, bin_len)
    json_bytes = _read_exact(stream, json_len, allow_eof_at_start=False)
    bin_bytes = _read_exact(stream, bin_len, allow_eof_at_start=False)
    envelope = json.loads((json_bytes or b"").decode("utf-8"))
    return envelope, bin_bytes or b""


def _check_caps(json_len: int, bin_len: int) -> None:
    if json_len > _MAX_JSON_LEN:
        raise ProtocolError(f"json part {json_len} bytes exceeds the {_MAX_JSON_LEN} byte cap")
    if bin_len > _MAX_BIN_LEN:
        raise ProtocolError(f"binary part {bin_len} bytes exceeds the {_MAX_BIN_LEN} byte cap")


def _read_exact(stream: BinaryIO, n: int, allow_eof_at_start: bool) -> Optional[bytes]:
    if n == 0:
        return b""
    chunks: List[bytes] = []
    remaining = n
    first = True
    while remaining > 0:
        chunk = stream.read(remaining)
        if not chunk:
            if first and allow_eof_at_start:
                return None
            raise ProtocolError("truncated frame: stream ended mid frame")
        chunks.append(chunk)
        remaining -= len(chunk)
        first = False
    return b"".join(chunks)
